src/keyboard/mod.rs

```diff
@@ -0,0 +1,258 @@
+            }
+            KeyState::Released => {
+                self.pressed_keys.retain(|&k| k != keycode);
+                xkb::KeyDirection::Up
+            }
+        };
+
+        // update state
+        let state_components = self.state.update_key(keycode, direction);
+        if state_components != 0 {
+            self.mods_state.update_with(&self.state);
+            true
+        } else {
+            false
+        }
+    }
+
+    fn serialize_modifiers(&self) -> (u32, u32, u32, u32) {
+        let mods_depressed = self.state.serialize_mods(xkb::STATE_MODS_DEPRESSED);
+        let mods_latched = self.state.serialize_mods(xkb::STATE_MODS_LATCHED);
+        let mods_locked = self.state.serialize_mods(xkb::STATE_MODS_LOCKED);
+        let layout_locked = self.state.serialize_layout(xkb::STATE_LAYOUT_LOCKED);
+
+        return (mods_depressed, mods_latched, mods_locked, layout_locked);
+    }
+
+    fn serialize_pressed_keys(&self) -> Vec<u8> {
+        let serialized = unsafe {
+            ::std::slice::from_raw_parts(self.pressed_keys.as_ptr() as *const u8,
+                                         self.pressed_keys.len() * 4)
+        };
+        serialized.into()
+    }
+}
+
+pub fn create_keyboard_handler() -> Result<KbdHandle, ()> {
+    let internal = KbdInternal::new()?;
+
+    // prepare a tempfile with the keymap, to send it to clients
+    // TODO: better error handling
+    let mut keymap_file = tempfile().unwrap();
+    let keymap_data = internal.keymap.get_as_string(xkb::KEYMAP_FORMAT_TEXT_V1);
+    keymap_file.write_all(keymap_data.as_bytes()).unwrap();
+    keymap_file.flush().unwrap();
+
+    Ok(KbdHandle {
+           internal: Arc::new((Mutex::new(internal), (keymap_file, keymap_data.as_bytes().len() as u32))),
+       })
+}
+
+/// An handle to a keyboard handler
+///
+/// It can be cloned and all clones manipulate the same internal state. Clones
+/// can also be sent across threads.
+///
+/// See module-level documentation for details of its use.
+#[derive(Clone)]
+pub struct KbdHandle {
+    internal: Arc<(Mutex<KbdInternal>, (::std::fs::File, u32))>,
+}
+
+impl KbdHandle {
+    /// Handle a keystroke
+    ///
+    /// All keystrokes from the input backend should be fed _in order_ to this method of the
+    /// keyboard handler. It will internally track the state of the keymap.
+    ///
+    /// The `filter` argument is expected to be a closure which will peek at the generated input
+    /// as interpreted by the keymap befor it is forwarded to the focused client. If this closure
+    /// returns false, the input will not be sent to the client. This mechanism can be used to
+    /// implement compositor-level key bindings for example.
+    ///
+    /// The module `smithay::keyboard::keysyms` exposes definitions of all possible keysyms
+    /// to be compared against. This includes non-characted keysyms, such as XF86 special keys.
+    pub fn input<F>(&self, keycode: u32, state: KeyState, serial: u32, filter: F)
```

Comment:
Actually, I was thinking this `filter` would often be a closure: it'd handle the logic of compositor keybindings and to do that would likely need to capture some environment.

As I see it, having it provided at this level (rather than set once at creation) makes this capture easier and avoids the need to add more synchronisation primitives.

That's what I had in mind. Maybe it's not realistic from an usability point of view?